python/kfserving/server.py

```diff
@@ -0,0 +1,141 @@
+            (r"/", LivenessHandler),
+            # Protocol Discovery API that returns the serving protocol supported by this server.
+            (r"/protocol", ProtocolHandler),
+            # Prometheus Metrics API that returns metrics for model servers
+            (r"/metrics", MetricsHandler, dict(models=self.models)),
+            # Model Health API returns 200 if model is ready to serve.
+            (r"/model/([a-zA-Z0-9_-]+)",
+             ModelHealthHandler, dict(models=self.models)),
+            # Predict API executes executes predict on input tensors
+            (r"/model/([a-zA-Z0-9_-]+)",
+             ModelPredictHandler, dict(models=self.models)),
+            # Optional Custom Predict Verb for Tensorflow compatibility
+            (r"/model/([a-zA-Z0-9_-]+):predict",
+             ModelPredictHandler, dict(models=self.models)),
+        ])
+
+        self._http_server.listen(self.http_port)
+        logging.info("Listening on port %s" % self.http_port)
+        tornado.ioloop.IOLoop.current().start()
+
+    def register_model(self, model):
+        if not model.name:
+            raise Exception("Failed to register model, invalid model name")
+        self.models[model.name] = model
+
+
+class LivenessHandler(tornado.web.RequestHandler):
+    def get(self):
+        self.write("Alive")
+
+
+class ProtocolHandler(tornado.web.RequestHandler):
+    def initialize(self, protocol):
+        self.protocol = protocol
+
+    def get(self):
+        self.write(self.protocol)
+
+
+class MetricsHandler(tornado.web.RequestHandler):
+    def get(self):
+        self.write("Not Implemented")
+
+
+class ModelHealthHandler(tornado.web.RequestHandler):
+    def initialize(self, models):
+        self.models = models
+
+    def get(self, name):
+        if name not in self.models:
+            raise tornado.web.HTTPError(
+                status_code=404,
+                reason="Model with name %s does not exist." % name
+            )
+
+        model = self.models[name]
+        self.write(json.dumps({
+            "name": model.name,
+            "ready": model.ready
+        }))
+
+
+class ModelPredictHandler(tornado.web.RequestHandler):
+    def initialize(self, models):
+        self.models = models
+
+    def post(self, name):
+        # TODO Add metrics
+        if name not in self.models:
+            raise tornado.web.HTTPError(
+                status_code=HTTPStatus.NOT_FOUND,
+                reason="Model with name %s does not exist." % name
+            )
+
+        model = self.models[name]
+        if not model.ready:
+            model.load()
+
+        try:
+            body = json.loads(self.request.body)
+        except json.decoder.JSONDecodeError as e:
+            raise tornado.web.HTTPError(
+                status_code=HTTPStatus.BAD_REQUEST,
+                reason="Unrecognized request format: %s" % e
+            )
+
+        if "instances" not in body:
```

Comment:
So we are creating a meta-schema that all protocols must adhere to? I thought we would be sending the protocol directly to the server?

Reply:
as discussed in meeting, this is TF API